lib/crowdsec.lua

```diff
@@ -456,15 +214,58 @@ function csmod.GetCaptchaBackendKey()
   return captcha.GetCaptchaBackendKey()
 end
 
-function csmod.SetupStream()
+local function SetupStream()
+  local function query_wrapper()
+    local err = stream:stream_query(
+      runtime.conf["API_URL"],
+      runtime.conf["REQUEST_TIMEOUT"],
+      REMEDIATION_API_KEY_HEADER,
+      runtime.conf["API_KEY"],
+      runtime.userAgent,
+      runtime.conf["SSL_VERIFY"],
+      runtime.conf["BOUNCING_ON_TYPE"]
+    )
+    if err ~=nil then
+      ngx.log(ngx.ERR, "Failed to query the stream: " .. err)
+      error("Failed to query the stream: " .. err)
+    end
+  end
   -- if it stream mode and startup start timer
   if runtime.conf["API_URL"] == "" then
     return
   end
+
+  ngx.log(ngx.DEBUG, "running timers: " .. tostring(ngx.timer.running_count()) .. " | pending timers: " .. tostring(ngx.timer.pending_count()))
+  local refreshing = stream.cache:get("refreshing")
+
+  if refreshing == true then
+    ngx.log(ngx.DEBUG, "another worker is refreshing the data, returning")
+    local ok, err = ngx.timer.at(runtime.conf["UPDATE_FREQUENCY"], SetupStream)
+    if not ok then
+      error("Failed to create the timer: " .. (err or "unknown"))
+    end
+    return
+  end
+
+
+  local last_refresh = stream.cache:get("last_refresh")
+  if last_refresh ~= nil then
+      -- local last_refresh_time = tonumber(last_refresh)
+      local now = ngx.time()
+      if now - last_refresh < runtime.conf["UPDATE_FREQUENCY"] then
+        ngx.log(ngx.DEBUG, "last refresh was less than " .. runtime.conf["UPDATE_FREQUENCY"] .. " seconds ago, returning")
+        local ok, err = ngx.timer.at(runtime.conf["UPDATE_FREQUENCY"], SetupStream)
+        if not ok then
+          error("Failed to create the timer: " .. (err or "unknown"))
+        end
+        return
+      end
+  end
+
   ngx.log(ngx.DEBUG, "timer started: " .. tostring(runtime.timer_started) .. " in worker " .. tostring(ngx.worker.id()))
-  if runtime.timer_started == false and runtime.conf["MODE"] == "stream" then
```

Comment:
why did the check on "stream" mode has been removed?

Reply:
It has not been removed you will find it in:
 ```
if runtime.conf["MODE"] == "stream" then
    ngx.log(ngx.INFO, "stream mode")
    SetupStream()
end
```

The metrics code doesn't depend on mode, so that's why it is not ran under a check with stream mode